py/desispec/scripts/trace_shifts.py

```diff
@@ -234,9 +234,17 @@ def fit_trace_shifts(image,args) :
         degxy=0
         degyx=0
         degyy=0
+    
+    nloops = 5
```

Comment:
I think the intended max number of iterations is supposed to be `args.degxx+args.degxy+args.degyx+args.degyy`, since each iteration through the loop decrements one or more of those until reaching the break condition on line 263:
```
            if degxx==0 and degxy==0 and degyx==0 and degyy==0 :
                break
```
hardcoding nloops=5 means that for larger values of `args.deg..` it could exit the loop earlier than originally intended.  i.e. I think your `and` vs. `or` bugfix below is the relevant change and should fix the known problems.  If you also want to add a belt-and-suspenders safety check to avoid the infinite loop that's ok, but then I think `nloops` should be larger than anything that is supposed to occur, and have nloops=0 trigger an exception rather than a standard `degxx, degxy, degyx, degyy = 0, 0, 0, 0` warning+proceed.

Reply:
Thanks for catching this too. I hadn't considered that we might ever fit very high-order polynomials but it's best to future-proof everything. As written by Julien, the total number of loops as we reduce the order of the polynomials down to zero should be
```
max(degxx, degyx) + max(degxy, degyy)
```
The code now checks that the number of loops does not exceed this value and throws a `RuntimeError` if it does.

Reply:
Just to demonstrate that the new code works, here are two functions that reproduce the old version of the loop and the new one:
```
def reduce_order_old(degxx=2, degyx=2, degxy=2, degyy=2):
    """Old loop from master branch.
    """
    n = 0
    while True:
        print(degxx, degyx, degxy, degyy)
        
        if degxy>0 and degyy>0 and degxy>degxx and degyy>degyx: # first along wavelength
            if degxy>0 : degxy-=1
            if degyy>0 : degyy-=1
        else :                                                  # then along fiber
            if degxx>0 : degxx-=1
            if degyx>0 : degyx-=1
            
        n += 1
        if n > 10:   # For testing, stop at 10 iterations
            break
            
def reduce_order_new(degxx=2, degyx=2, degxy=2, degyy=2):
    """Updated loop from master branch.
    """
    nloops = max(degxx, degyx) + max(degxy, degyy)
    n = 0
    while True:
        print(degxx, degyx, degxy, degyy)
        
        if (degxy>0 or degyy>0) and (degxy>degxx or degyy>degyx): # first along wavelength
            if degxy>0 : degxy-=1
            if degyy>0 : degyy-=1
        else :                                                    # then along fiber
            if degxx>0 : degxx-=1
            if degyx>0 : degyx-=1
            
        if nloops == 0:
            break

        n += 1
        if n > nloops:
            break
```

When running with `degxx=degyx=degxy=degyy=2` (the default arguments) both functions return the same values:
```
reduce_order_old(2,2,2,2)     reduce_order_new(2,2,2,2)
2 2 2 2                       2 2 2 2
1 1 2 2                       1 1 2 2
1 1 1 1                       1 1 1 1
0 0 1 1                       0 0 1 1
0 0 0 0                       0 0 0 0
```
But when `degyy=0` the original loop does not properly decrease to 0 for all orders:
```
reduce_order_old(2,2,2,2)     reduce_order_new(2,2,2,2)
2 2 2 0                       2 2 2 0
1 1 2 0                       1 1 2 0
0 0 2 0                       1 1 1 0
0 0 2 0                       0 0 1 0
0 0 2 0                       0 0 0 0
0 0 2 0
...
```